src/CommandingOfficers/Ave.java

```diff
@@ -620,38 +644,76 @@ protected void perform(MapMaster gameMap)
 
       // Keep track of any tiles that change.
       ArrayList<MapChangeEvent.EnvironmentAssignment> tileChanges = new ArrayList<MapChangeEvent.EnvironmentAssignment>();
-      ArrayList<Unit> victims = new ArrayList<Unit>();
 
       // Change terrain to snow around each of Ave's units and buildings, and damage trees and enemies.
-      Set<XYCoord> affectedTiles = Utils.findLocationsNearProperties(gameMap, Ave, OBLIDO_RANGE);
-      affectedTiles.addAll(Utils.findLocationsNearUnits(gameMap, Ave, OBLIDO_RANGE));
+      Set<XYCoord> affectedTiles = getTilesInRange(gameMap);
 
       // Smash things. Don't add snow though.
       for( XYCoord coord : affectedTiles )
       {
         // Destroy any forests. Big hail, man.
-        Location loc = gameMap.getLocation(coord);
-        Environment tileEnvi = loc.getEnvironment();
+        Environment tileEnvi = gameMap.getEnvironment(coord);
         if(tileEnvi.terrainType == TerrainType.FOREST)
         {
-          tileChanges.add(new MapChangeEvent.EnvironmentAssignment(coord, Environment.getTile(TerrainType.GRASS, loc.getEnvironment().weatherType), 1));
-        }
-
-        // Damage each enemy nearby.
-        if( null != loc.getResident() && myCommander.isEnemy(loc.getResident().CO) )
-        {
-          victims.add(loc.getResident());
+          tileChanges.add(new MapChangeEvent.EnvironmentAssignment(coord, Environment.getTile(TerrainType.GRASS, tileEnvi.weatherType), 1));
         }
       }
 
-      GameEvent damage = new MassDamageEvent(victims, 2, false);
+      GameEvent damage = new MassDamageEvent(findVictims(gameMap, affectedTiles), OBLIDO_DAMAGE, false);
       damage.performEvent(gameMap);
       GameEventListener.publishEvent(damage);
 
       GameEvent tileChange = new MapChangeEvent(tileChanges);
       tileChange.performEvent(gameMap);
       GameEventListener.publishEvent(tileChange);
     }
+
+    @Override
+    public Collection<DamagePopup> getDamagePopups(GameMap gameMap)
+    {
+      ArrayList<DamagePopup> output = new ArrayList<DamagePopup>();
+      Set<XYCoord> affectedTiles = getTilesInRange(gameMap);
+
+      for( Unit victim : findVictims(gameMap, affectedTiles) )
+      {
+        XYCoord coord = new XYCoord(victim.x, victim.y);
+        // BAMF takes priority over damage
+        if( gameMap.getEnvironment(coord).terrainType != TerrainType.FOREST )
+          output.add(new DamagePopup(
+                         coord,
+                         myCommander.myColor,
+                         Math.min(victim.getHP()-1, OBLIDO_DAMAGE)*10 + "%"));
+      }
+      for( XYCoord coord : affectedTiles )
+        if( gameMap.getEnvironment(coord).terrainType == TerrainType.FOREST )
+          output.add(new DamagePopup(
+                         coord,
+                         myCommander.myColor,
+                         "BAMF"));
```

Comment:
BAMF? You choose this acronym for a particular reason?

Reply:
Is it an acronym?

Reply:
[*cough*](https://www.urbandictionary.com/define.php?term=BAMF)

Reply:
I was living in the world of comic books, where "BAMF" is what happens when Batman punches you.

Reply:
Haha, I kinda figured. I move we find a different onomotopeia, or pick something descriptive like "RAZE" or whatever.